pkg/traversal/traversal.go

```diff
@@ -0,0 +1,133 @@
+	"github.com/openshift/must-gather-clean/pkg/omitter"
+	"github.com/openshift/must-gather-clean/pkg/output"
+)
+
+type FileWalker struct {
+	input       string
+	obfuscators []obfuscator.Obfuscator
+	writer      output.Outputter
+	omitters    []omitter.Omitter
+}
+
+// NewFileWalker returns a FileWalker. It ensures that the input directory exists and is readable.
+func NewFileWalker(inputDir string, writer output.Outputter, obfuscators []obfuscator.Obfuscator, omitters []omitter.Omitter) (*FileWalker, error) {
+	fileInfo, err := os.Stat(inputDir)
+	if err != nil {
+		return nil, fmt.Errorf("cannot access input directory %s: %w", inputDir, err)
+	}
+	if !fileInfo.IsDir() {
+		return nil, fmt.Errorf("input path %s is not a directory", inputDir)
+	}
+	inputPath, err := filepath.Abs(inputDir)
+	if err != nil {
+		return nil, err
+	}
+	return &FileWalker{input: inputPath, obfuscators: obfuscators, writer: writer, omitters: omitters}, nil
+}
+
+// Traverse should be called to start processing the input directory.
+func (w *FileWalker) Traverse() error {
+	return w.processDir(w.input, "")
+}
+
+func (w *FileWalker) processDir(inputDirName string, outputDirName string) error {
+	// list all the entities in the directory
+	entries, err := os.ReadDir(inputDirName)
+	if err != nil {
+		return err
+	}
+	for _, e := range entries {
+		fileInfo, err := e.Info()
+		if err != nil {
+			return err
+		}
+
+		// If a file is a symlink then ignore it
+		if fileInfo.Mode()&os.ModeSymlink != 0 {
+			continue
+		}
+
+		if e.IsDir() {
+			name := e.Name()
+			// obfuscate the directory name
+			for _, o := range w.obfuscators {
+				name = o.FileName(name)
+			}
+			// create the directory
+			err := w.writer.MakeDir(outputDirName, name, fileInfo.Mode().Perm())
+			if err != nil {
+				return err
+			}
+			childDirOutput := path.Join(outputDirName, name)
+			childDirInput := path.Join(inputDirName, e.Name())
+			err = w.processDir(childDirInput, childDirOutput)
+			if err != nil {
+				return err
+			}
+		} else {
+			leafPath := path.Join(inputDirName, e.Name())
+
+			var omit bool
+			// verify if the file should be omitted
+			for _, o := range w.omitters {
+				omit, err = o.File(e.Name(), leafPath)
+				if err != nil {
+					return fmt.Errorf("failed to determine if %s should be omitted based on name: %w", leafPath, err)
+				}
+				if omit {
+					break
+				}
+				omit, err = o.Contents(leafPath)
+				if err != nil {
+					return fmt.Errorf("failed to determine if %s should be omitted based on contents: %w", leafPath, err)
+				}
+				if omit {
+					break
+				}
+			}
+			// If the file should be omitted then stop processing.
+			if omit {
+				continue
+			}
+			return func() error {
+				writeCloser, writer, err := w.writer.Writer(outputDirName, e.Name(), fileInfo.Mode())
+				defer func() {
+					if err := writeCloser(); err != nil {
+						fmt.Printf("failed to successfully write file %s: %v", path.Join(outputDirName, e.Name()), err)
```

Comment:
can you please use `filepath` instead of path? it's known to not join properly on windows for example (due to the backslashes and escaping).

Reply:
just some background reading: https://stackoverflow.com/questions/39181790/whats-the-difference-between-path-path-filepath-packages-in-go